rstudio/shiny-server
# Fix issues with log_as_user

* The mkdir.sh script attempted to chown a directory. I don't know
  if this could ever work? If you're running mkdir.sh, you're not
  root, and isn't chown something generally reserved for root?
  Anyway, we don't do this anymore.

* log_file_mode wasn't respected if log_as_user was true; now it
  is. We do this in the same mkdir.sh script, and for this reason
  it is now called create-log.sh instead.

### TODO

- [x] Default is still a number, make it a string.
- [x] Add unit test that ensures fs.chown, etc. take mode as a String

### Testing notes

@shalutiwari This fixes issues around NFS root_squash. The gist of it is that normally Shiny Server runs as root, and assumes that the root user can create log files, and set ownership and permissions on them. That turns out to not be true in the case of NFS shared directories; NFS by default has a feature called root_squash that downgrades a client machine's root account to a "nobody" account instead. (It's possible, but *extremely* dangerous, to run an NFS shared directory without root_squash.) In cases where Shiny Server's `log_dir` is on an NFS root_squash shared directory, root doesn't have enough permissions to create new log files, but users do. So `log_as_user true;` says to create and set permissions on log files as the `shiny` (or whatever) user, not as root.

This PR includes changes to the `vagrant/nfs` directory. You will need to use this vagrant setup to replicate the issue, unless you want to set up your own NFS server.

Make sure you are using vagrant 2.x, not 1.x. Otherwise it won't find the Ubuntu 14.04 base image, probably.

1. `cd vagrant/nfs`
2. `vagrant up nfs sso` (creates two separate VMs; `nfs` will serve up a shared folder via NFS, `sso` will mount the shared folder and run Shiny Server)
3. `vagrant ssh nfs` to log into the file server. Verify that `/home/shiny/ShinyApps/log` is either empty or doesn't exist.
4. Visit http://192.168.42.101:3838/shiny/01_hello/ in a browser, it should work.
5. `ls -l /home/shiny/ShinyApps/log` and `ls -ld /home/shiny/ShinyApps/log`; verify that both the `log` folder and any log files inside are owned by `shiny` user.

---

6. In a separate terminal, go to the `vagrant/nfs` directory and `vagrant ssh sso` to log into the Shiny Server VM.
7. Add a line to `/etc/shiny-server/shiny-server.conf`: `log_file_mode 0600;`. You can put it right below `log_as_user true;`.
8. `sudo systemctl restart shiny-server`
9. Visit http://192.168.42.101:3838/shiny/01_hello/ again.
10. `ls -l /home/shiny/ShinyApps/log`, there should be a new log file created with `rw-------` permissions.

---

It would also be a good idea to test using `log_file_mode` when `log_as_user false;` (the default) is used. You shouldn't use the above vagrant setup for that; you can just use whatever setup you normally use to test Shiny Server.
